core/src/smartcontracts/wasm.rs

```diff
@@ -459,6 +465,56 @@ impl<'wrld> Runtime<'wrld> {
         Ok(())
     }
 
+    /// Host-defined function which logs the given string. When this function
+    /// is called, the module serializes the string to linear memory and
+    /// provides log level, offset and length as parameters
+    ///
+    /// # Warning
+    ///
+    /// This function doesn't take ownership of the provided
+    /// allocation
+    ///
+    /// # Errors
+    ///
+    /// If log level or string decoding fails
+    fn log(
+        mut caller: Caller<State>,
+        log_level: u32,
+        offset: WasmUsize,
+        len: WasmUsize,
+    ) -> Result<(), Trap> {
+        let error_msg = Trap::new(format!("{}: not a valid log level", log_level));
```

Comment:
This is super picky, but ideally you could write this:

```suggestion
        let construct_error_msg = || Trap::new(format!("{}: not a valid log level", log_level));
```

This will reduce extra allocation on happy path which is the most common